File: linkedlist/javacode/LinkedlistLoops.java
package javacode;

class LinkedlistLoops extends Linkedlist {
  // Loop Distance Using Floyd's Algorithm and Additional Pointer
  public void loopLength() {
    Node slow = head;
    Node fast = head;
    Node startPtr = head;
    int loopSize = 1;
    while (fast.next != null && fast.next.next != null) {
      slow = slow.next;
      fast = fast.next.next;
      if (slow == fast) {
        break;
      }
    }
    while (true) {
      if (slow != startPtr) {
        slow = slow.next;
        startPtr = startPtr.next;
      } else if (slow == startPtr) {
        slow = slow.next;
        while (startPtr != slow) {
          loopSize++;
          slow = slow.next;
        }
        break;
      }
    }
    System.out.println("Length of the loop is: " + loopSize);
  }

  // Starting point of loop using Floyd's Algorithm
  public void loopStart() {
    Node slow = head;
    Node fast = head;
    Node startPtr = head;

    while (fast.next != null && fast.next.next != null) {
      slow = slow.next;
      fast = fast.next.next;
      if (slow == fast) {
        while (slow != startPtr) {
          slow = slow.next;
          startPtr = startPtr.next;
        }
        break;
      }
    }
    System.out.println("Starting point of loop is: " + startPtr.val);
  }

  // Detect Loop Using Slow and Fast Pointer or Floyd's algorithm
  public void detectLoop() {
    Node slow = head;
    Node fast = head;

    while (fast.next != null && fast.next.next != null) {
      slow = slow.next;
      fast = fast.next.next;
      if (slow == fast) {
        System.out.println("Loop Exist");
        return;
      }
    }
    System.out.println("Loop doesn't exist");
  }

  // Middle ElementS using slow and fast pointer
  public void middleElem() {
    Node slow = head;
    Node fast = head;

    for (;;) {
      if (fast.next != null && fast.next.next != null) {
        slow = slow.next;
        fast = fast.next.next;
      } else {
        break;
      }
    }
    if (fast.next != null) {
      slow = slow.next;
    }
    System.out.println("Middle element of list is: " + slow.val);
  }

  // Create loop in LinkedlistS
  public void createCycle(int index) {
    Node temp = head;
    Node indexNode = head;
    int i = 0;
    while (temp != null) {
      if (i != index) {
        i++;
        indexNode = indexNode.next;
      }
      if (temp.next == null) {
        temp.next = indexNode;
        break;
      }
      temp = temp.next;
    }
  }
}

File: linkedlist/singly/linkedlistloops.go
package main

// Length of loop
func getLoopLength(head **node) int {
	slow := *head
	fast := *head
	// loop detection
	for fast.next != nil && fast.next.next != nil {
		slow = slow.next
		fast = fast.next.next
		if slow == fast {
			break
		}
	}
	// starting point of loop
	entry := *head
	for {
		if entry == slow {
			break
		} else {
			entry = entry.next
			slow = slow.next
		}
	}
	//length of the loop
	size := 0
	for {
		slow = slow.next
		size++
		if entry == slow {
			break
		}
	}
	return size
}

// Starting point of loop
// Also called as Floyd's Algorithm
func getloopstartnode(head **node) *node {
	slow := *head
	fast := *head
	// loop detection
	for fast.next != nil && fast.next.next != nil {
		slow = slow.next
		fast = fast.next.next
		if slow == fast {
			break
		}
	}
	// to find starting point of loop
	entry := *head
	for {
		if entry == slow {
			break
		} else {
			entry = entry.next
			slow = slow.next
		}
	}
	return entry
}

// Detect loop
func detectLoop(head **node) bool {
	slow := *head
	fast := *head

	for fast.next != nil && fast.next.next != nil {
		slow = slow.next
		fast = fast.next.next
		if slow == fast {
			return true
		}
	}

	return false
}

// Middle Element in Linkedlist
func getMiddleElem(head **node) int {
	if head == nil {
		return -1
	}
	if (*head).next == nil {
		return (*head).val
	}
	slow := *head
	fast := *head
	for fast.next != nil && fast.next.next != nil {
		slow = slow.next
		fast = fast.next.next
	}
	if fast.next != nil {
		slow = slow.next
	}
	return slow.val
}
